service/awsconfig/v10/ebs/ebs.go

```diff
@@ -0,0 +1,191 @@
+package ebs
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/aws-operator/service/awsconfig/v10/key"
+)
+
+const (
+	cloudProviderClusterTagValue        = "owned"
+	cloudProviderPersistentVolumeTagKey = "kubernetes.io/created-for/pv/name"
+	nameTagKey                          = "Name"
+)
+
+type Config struct {
+	Client EC2Client
+	Logger micrologger.Logger
+}
+
+type EBS struct {
+	client EC2Client
+	logger micrologger.Logger
+}
+
+func New(config Config) (*EBS, error) {
+	if config.Client == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Client must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	c := &EBS{
+		client: config.Client,
+		logger: config.Logger,
+	}
+
+	return c, nil
+}
+
+// DeleteVolume deletes an EBS volume with retry logic.
+func (e *EBS) DeleteVolume(ctx context.Context, volumeID string) error {
+	e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting ebs volume %s", volumeID))
+
+	deleteOperation := func() error {
+		_, err := e.client.DeleteVolume(&ec2.DeleteVolumeInput{
+			VolumeId: aws.String(volumeID),
+		})
+		if IsVolumeNotFound(err) {
+			// Fall through.
+			return nil
+		}
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		return nil
+	}
+	deleteNotify := func(err error, delay time.Duration) {
+		e.logger.LogCtx(ctx, "level", "error", fmt.Sprintf("deleting ebs volume failed, retrying with delay %.0fm%.0fs: '%#v'", delay.Minutes(), delay.Seconds(), err))
+	}
+	deleteBackoff := &backoff.ExponentialBackOff{
+		InitialInterval:     backoff.DefaultInitialInterval,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          backoff.DefaultMultiplier,
+		MaxInterval:         backoff.DefaultMaxInterval,
+		MaxElapsedTime:      3 * time.Minute,
```

Comment:
Why do we want to have it this long lived? Is there a story behind this decision here? My thinking is that we should maybe try three times and then let other objects reconcile to not block resources for too long. We might be better off with this on the next resync period? Or is a retry in here even necessary?

Reply:
I agree this is too long. I want to reduce this once I've confirmed the deletion is working. I'll do this as a later change.

Reply:
I'd like to keep the retry but I've reduced it to 30 seconds. Also the resource now consumes errors so the deletion logic in following resources is executed.